pyccel/parser/semantic.py

```diff
@@ -2131,27 +2131,32 @@ def _visit_Assign(self, expr, **settings):
             else:
 
                 # TODO check types from FunctionDef
-
                 master = macro.master
                 name = _get_name(master.name)
-
-                # all terms in lhs must be already declared and available
-                # the namespace
-                # TODO improve
+                results = []
+                args = [self._visit(i, **settings) for i in rhs.args]
+                args_names = [arg.name for arg in args if isinstance(arg, Variable)]
 
                 if not sympy_iterable(lhs):
                     lhs = [lhs]
+                results_shapes = macro.get_results_shapes(args)
+                for m_result, shape, result in zip(macro.results, results_shapes, lhs):
+                    if m_result in macro.master_arguments and not result in args_names:
+                        index = macro.master_arguments.index(m_result)
+                        d_result = self._infere_type(master.arguments[index])
```

Comment:
Why can you not just do `d_result = self._infere_type(m_result)`? I'm not sure what `index` is adding here

Reply:
`m_result` is coming from the macro, it is still of type `PyccelSymbol`, that's why i need to get the index of m_result, to get the according argument from the `FunctionDef` `master`